docs/source/traits_user_manual/observing.rst

```diff
@@ -0,0 +1,538 @@
+   * - |list_items|
+     - For observing items in a list.
+   * - |set_items|
+     - For observing items in a set.
+   * - |match|
+     - For observing traits satisfying a user-defined filter.
+   * - |parse|
+     - For parsing a string written in Traits domain specific language into an
+       expression object.
+
+Users can combine complex expressions using |ObserverExpression.then| or
+Python's bitwise-or (`|`) operation.
+
+Most of these functions will have a **notify** argument for setting whether
+notifications should be fired for changes.
+
+.. rubric:: Example expressions
+
+.. list-table::
+   :widths: 15 25
+   :header-rows: 1
+
+   * - Expression
+     - Meaning
+   * - *trait("attr1")*
+     - Matches a trait named *attr1* on an object and notifies for changes.
+   * - *trait("attr1", optional=True)*
+     - Matches a trait named *attr1* if it is defined. Ignore if it is not
+       defined.
+   * - *trait("attr1", notify=False).trait("attr2")*
+     - Matches a trait named *attr2* on an object referenced by a trait named
+       *attr1* on the current object. Changes to *attr2* will trigger
+       notifications, while changes to *attr1* do not.
+   * - *trait("foo").list_items().list_items().trait("value")*
+     - Matches the *value* trait on an item of a nested list in another list
+       *foo*. Assignment changes to *foo*, mutations to the lists or changes
+       to *value* will trigger notifications.
+   * - *metadata("updated")*
+     - Matches any trait on the current that has a metadata attribute named
+       *updated*. Changes on those traits will trigger notifications.
+   * - *trait("foo") | trait("bar")*
+     - Matches trait named *foo* or *bar* on the current object. Changes on
+       *foo* or *bar* will trigger notifications.
+   * - *trait("foo").then(trait("bar") | trait("baz"))*
+     - Matches *foo.bar* or *foo.baz* on the current object. Changes on *foo*,
+       *foo.bar* or *foo.baz* will trigger notifications.
+   * - *trait("foo").match(lambda n, t: True)*
+     - Matches any traits on *foo* on the current object. Changes on *foo* or
+       the nested attributes will trigger notifications.
+
+.. rubric:: Extend an expression in text
+
+Using the |parse| function, one can extend an expression in text
+with additional features supported by |ObserverExpression|. For example::
+
+    parse("foo.bar").match(lambda name, trait: name.startswith("my_"))
+
+will observe traits with a prefix "my\_" on *foo.bar* on the current object.
+
+
+.. _observe-handler:
+
+Notification Handler
+--------------------
+
+As soon as a change has occurred, the **handler** callable provided will be
+invoked as long as the change is observed for notifications. The handler will
+be given information about the change, and it must have the following
+signature:
+
+- handler(*event*)
+
+where the *event* parameter is an object representing the change observed.
+The type of *event* depends on the context of the change.
```

Comment:
We should mention here any guarantees that we make about the common API, eg. if `object` is always present.